rotary/rules/polish.smk

```diff
@@ -15,37 +15,120 @@ rule prepare_medaka_polish_input:
+    output:
+        contig_polished=temp('{sample}/{step}/medaka/results/{sample}_{contig}.hd5')
+    conda:
+        "../envs/medaka.yaml"
+    log:
+        "{sample}/logs/{step}/medaka_{sample}_{contig}.log"
+    benchmark:
+        "{sample}/benchmarks/{step}/medaka_{sample}_{contig}.txt"
     params:
         medaka_model=config.get("medaka_model"),
-        batch_size=config.get("medaka_batch_size")
+        batch_size=config.get("medaka_batch_size"),
+        medaka_results_dir='{sample}/{step}/medaka/results',
+    threads:
+        2
+    shell:
+        """
+        medaka consensus {input.calls_to_draft_bam} \
+          {params.medaka_results_dir}/{wildcards.sample}_{wildcards.contig}.hd5 \
+          --model {params.medaka_model} \
+          --batch {params.batch_size} \
+          --threads {threads} \
+          --region {wildcards.contig} > {log} 2>&1
+        """
+
+def aggregate_medaka_polished_contigs(wildcards):
+    """
+    Callback function that generates a list contig HDF5 files that will be needed for rule stitch_medaka.
+
+    :param wildcards: These are the wildcards present in rule stich_medaka.
+    :return: HDF5 files to be generated for each contig by multiple executions of rule polish_contig_medaka.
+    """
+    # Force execution of checkpoint generate_contig_files.
+    # Passes wildcards from rule stitch_medaka to generate_contig_files.
+    # Execution will generate a contig manifest file and revaluate the DAG.
+    contig_manifest_path = checkpoints.generate_contig_manifest.get(**wildcards).output[0]
+
+    # Open the resulting contig manifest file and read in the contig names.
+    with contig_manifest_path.open() as contig_file:
+        contigs_names = [line.strip() for line in contig_file]
+
+    # Returns the expected paths to the per-contig HDF5 polishing files.
+    return expand("{sample}/{step}/medaka/results/{sample}_{contig}.hd5",sample=wildcards.sample,
+        step=wildcards.step,contig=contigs_names)
+
+rule stitch_medaka:
+    input:
+        hdf5s=aggregate_medaka_polished_contigs,
+        draft_fasta_dir="{sample}/polish/medaka_input/",
+        draft_fasta="{sample}/polish/medaka_input/{sample}_input.fasta"
+    output:
+        "{sample}/{step}/medaka/{sample}_consensus.fasta"
+    conda:
+        "../envs/medaka.yaml"
+    log:
+        "{sample}/logs/{step}/medaka_stich.log"
+    benchmark:
+        "{sample}/benchmarks/{step}/medaka_stich.txt"
     threads:
         config.get("threads",1)
+    params:
+        medaka_model=config.get("medaka_model"),
+        batch_size=config.get("medaka_batch_size")
     shell:
         """
-        medaka_consensus -i {input.qc_long_reads} -d {input.contigs} -o {output.dir} \
-          -m {params.medaka_model} -t {threads} -b {params.batch_size} > {log} 2>&1
-        mv {output.dir}/consensus.fasta {output.contigs}
+        medaka stitch --threads {threads} {input.hdf5s} {input.draft_fasta} {output} > {log} 2>&1
```

Comment:
I think this is an inherent issue with the medaka stitch CLI itself, but note that, if the assembly file contains a huge number of contigs, the shell will error out with an "argument list too long" error due to the large number of HDF5 files (e.g., see https://stackoverflow.com/a/19355351)

Reply:
So, if you look at the code for `medaka_concensus,` it's a shell script that contains all the steps in the medaka pipeline. It is called `medaka stitchch,` so it should run into the same issue.

https://github.com/nanoporetech/medaka/blob/master/scripts/medaka_consensus

Reply:
So I think we should leave this. But thanks for the heads up about a potential issue.

Reply:
@jmtsuji I think we can leave this agreed?

Reply:
@LeeBergstrand Thanks for the additional check of the `medaka_consensus` code -- interesting that this is a shell script! I agree with you, let's leave our code as-is for now.

The possible case I thought of when our code might break is if someone wanted to use our pipeline to assemble a soil metagenome or something. They might have hundreds of thousands of contigs. In the `medaka_consensus` shell script, I think all those contigs would be kept in one HDF5 file rather than being split up into 1 contig per HDF5, because that script is not parallelized (in my understanding). So our code might break in this case, whereas the `medaka_consensus` script might still work. But it's probably OK for us to face that issue later if/when it comes up. We might need to batch the HDF5's differently or stitch in several batches or something.

Reply:
@jmtsuji Yes 1) we would have to evaluate how many contigs would cause this failure to occur. I suspect that it's probably much higher than expected. 2) We would probably have to add other steps and changes to support metagenomes explicitly. So we can think about that issue then.

Reply:
Okay, sounds good. Let's consider this more later, if/when needed.